### PR TITLE
feat: façade converter handles various special cases for Dart.

### PR DESCRIPTION
**Pull request by mprobst**
- Bundles CONST_EXPR and FORWARD_REF into one place
- Special cases a bunch of Array methods, allowing more idiomatic TypeScript
  & Dart code
- Code structure should allow straight forward future additions.

### CODE REVIEW

**Comment by mprobst** [comment]
/CC @mhevery 

Partially handles angular/angular#2231.

File: lib/call.ts [hunk 1]
```diff
@@ -26,21 +27,11 @@ class CallTranspiler extends base.TranspilerStep {
...
-        if (this.isInsideConstExpr(node)) {
-          this.emit('const');
-        }
-        return false;
       case ts.SyntaxKind.CallExpression:
         var callExpr = <ts.CallExpression>node;
-        if (!this.maybeHandleSuperCall(callExpr)) {
-          this.visitCall(callExpr);
-        }
+        if (this.fc.handleCall(callExpr)) break;
```
Comment (alexeagle) [inline lib/call.ts]: I like the maybe* naming convention, want to use it here too?

Comment (alexeagle) [inline lib/call.ts]: opt: could use boolean OR rather than break, eg.  `this.fc.handle || this.maybeHandleSuper || this.visitCall`

File: lib/expression.ts [hunk 2]
```diff
@@ -2,9 +2,10 @@
 import ts = require('typescript');
 import base = require('./base');
 import ts2dart = require('./main');
```
Comment (alexeagle) [inline lib/expression.ts]: these require statements should be ES6 module instead?

File: lib/facade_converter.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    var qnSub = fileSubs[qn];
+    if (!qnSub) return false;
+
+    qnSub(c, context);
+    return true;
+  }
+
+  private subs: ts.Map<ts.Map<FacadeHandler>> = {
+    'lib': {
+      'Array.push': (c: ts.CallExpression, context: ts.Expression) => {
```
Comment (alexeagle) [inline lib/facade_converter.ts]: can type inference work here, so you could just `(c, context) => {}`
would make this structure more readable.

File: lib/facade_converter.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/// <reference path='../node_modules/typescript/bin/typescript.d.ts' />
+import base = require('./base');
+import ts = require('typescript');
+import ts2dart = require('./main');
+
+type FacadeHandler = (c: ts.CallExpression, context: ts.Expression) => void;
+
+export class FacadeConverter extends base.TranspilerStep {
```
Comment (alexeagle) [inline lib/facade_converter.ts]: seems unintuitive that this is a base.TranspilerStep but it doesn't handle any nodes on its own, instead is just a helper. Is it really a subtype?

File: lib/facade_converter.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    super(transpiler);
+    this.forbiddenNames = {};
+    for (var fileName in this.subs) {
+      Object.keys(this.subs[fileName])
+          .map((fnName) => fnName.substring(fnName.lastIndexOf('.') + 1))
+          .forEach((fnName) => this.forbiddenNames[fnName] = true);
+    }
+  }
+
+  setTypeChecker(tc: ts.TypeChecker) { this.tc = tc; }
```
Comment (alexeagle) [inline lib/facade_converter.ts]: still don't like ES6 setter syntax? :)

File: lib/main.ts [hunk 6]
```diff
@@ -74,6 +86,9 @@ export class Transpiler {
     }
     var destinationRoot = destination || this.options.basePath || '';
     var program = ts.createProgram(fileNames, OPTIONS, host);
+    if (this.options.translateBuiltins) {
+      this.fc.setTypeChecker(program.getTypeChecker());
```
Comment (alexeagle) [inline lib/main.ts]: why not make the typeChecker a constructor param on FacadeConverter?

File: test/facade_converter_test.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      new<K, V>(l: List<any>): Map<K, V>;
+      prototype: Map<any, any>;
+    };`;
+
+var langDeclarations = `
+  export declare function CONST_EXPR<T>(x: T): T;
+  export declare function FORWARD_REF<T>(x: T): T;
+  `;
+
+describe('collection façade', () => {
```
Comment (alexeagle) [inline test/facade_converter_test.ts]: you had to put it somewhere...

File: test/facade_converter_test.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        .to.equal(' const x = const { "one" : 1 } ;');
+  });
+
+  it('translates FORWARD_REF(() => T) to T', () => {
+    expectWithTypes('import {FORWARD_REF} from "angular2/src/facade/lang";\n' +
+                    'var x = FORWARD_REF(() => SomeType);')
+        .to.equal(' var x = SomeType ;');
+    chai.expect(() => expectWithTypes('import {FORWARD_REF} from "angular2/src/facade/lang";\n' +
+                                      'FORWARD_REF(1)'))
+        .to.throw(/only arrow functions/);
```
Comment (alexeagle) [inline test/facade_converter_test.ts]: ooh that is very pretty syntax for less brittle exception assertion

File: lib/call.ts (hunk 1 above)
Comment (mprobst) [inline lib/call.ts]: Done re maybe convention.
Regarding the ||, I think this is not much more code and reads a bit cleaner to me.

File: lib/expression.ts (hunk 2 above)
Comment (mprobst) [inline lib/expression.ts]: Ultimately, yes, but let's do that in another commit.

File: lib/facade_converter.ts (hunk 4 above)
Comment (mprobst) [inline lib/facade_converter.ts]: As discussed, renamed this to TranspilerBase.

File: lib/facade_converter.ts (hunk 5 above)
Comment (mprobst) [inline lib/facade_converter.ts]: Still don't like setters, ES6 or not ;-)

File: lib/main.ts (hunk 6 above)
Comment (mprobst) [inline lib/main.ts]: Yeah, the transpiler steps are created before creating the program (they can be used multiple times over different programs), and I can only get the TypeChecker from the program, but not from the individual nodes we're visiting. The alternative would be passing down the TypeChecker with the program nodes, but that seems overkill.

File: test/test_support.ts [hunk 9]
```diff
@@ -1,31 +1,57 @@
...
-      if (sourceName === fileName) {
-        return ts.createSourceFile(sourceName, contents, compilerOptions.target, true);
+      if (nameToContent.hasOwnProperty(sourceName)) {
+        return ts.createSourceFile(sourceName, nameToContent[sourceName], compilerOptions.target,
+                                   true);
       }
       if (sourceName === 'lib.d.ts') {
-        return ts.createSourceFile(sourceName, '', compilerOptions.target, true);
+        if (!libSourceFile) {
+          // Cache to avoid excessive test times.
```
Comment (alexeagle) [inline test/test_support.ts]: potential for test isolation failures? are the cache contents immutable?

**Comment by alexeagle** [comment]
LGTM when Travis says so

File: test/test_support.ts (hunk 9 above)
Comment (mprobst) [inline test/test_support.ts]: Yes, everything in tsc is immutable AFAIK.
